File: GroupA/04_NQueens/nQueens.cpp
#include <iostream>
#include <vector>

using namespace std;

void printSolution(const vector<vector<int> >& board) {
    int N = board.size();
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cout << (board[i][j] ? "Q " : ". ");
        }
        cout << endl;
    }
}

bool isSafe(const vector<vector<int> >& board, int row, int col) {
    int N = board.size();
    int i, j;

    for (j = 0; j < col; ++j) {
        if (board[row][j])
            return false;
    }

    for (i = row, j = col; i >= 0 && j >= 0; --i, --j) {
        if (board[i][j])
            return false;
    }

    for (i = row, j = col; i < N && j >= 0; ++i, --j) {
        if (board[i][j])
            return false;
    }

    return true;
}

bool solveNQUtil(vector<vector<int> >& board, int col) {
    int N = board.size();

    if (col == N) {
        return true;
    }

    for (int row = 0; row < N; ++row) {
        if (isSafe(board, row, col)) {
            board[row][col] = 1;

            if (solveNQUtil(board, col + 1)) {
                return true;
            }

            board[row][col] = 0;
        }
    }

    return false;
}

void solveNQ(int N, int firstQueenRow) {
    if (firstQueenRow < 0 || firstQueenRow >= N) {
        cout << "Invalid row for first queen." << endl;
        return;
    }

    vector<vector<int> > board(N, vector<int>(N, 0));

    board[firstQueenRow][0] = 1;

    if (solveNQUtil(board, 1)) {
        cout << "Solution found for N = " << N << " with first queen at (row=" << firstQueenRow << ", col=0):\n";
        printSolution(board);
    } else {
        cout << "No solution exists for N = " << N << " with first queen at (row=" << firstQueenRow << ", col=0)." << endl;
    }
}


int main() {
    int N = 8;
    int firstQueenRow = 0;

    solveNQ(N, firstQueenRow);
    
    cout << "\n--- Trying another combination ---\n";
    
    N = 4;
    firstQueenRow = 1;
    solveNQ(N, firstQueenRow);
    
    return 0;
}

File: GroupA/04_NQueens/nQueens.py
"""
    # NQueens but harder.
    In this problem the user has to be able to place a Queen which CANNOT be moved. 
    We have to find a way around it. 
    If a solution doesn't exist we tell that to the user.
    Basically we have to improve the `isSafe()` function of the regular backtracking NQueens problem, such that it detects both ways on the board for threats.
"""

def isSafe(pos:tuple, board: list)->bool:
    """Determines if the given position is safe to place a queen or not."""
    
    ROW = COL = len(board)
    row, col = pos

    # Check ROW 
    for r in range(ROW):
        if r != row and board[r][col].lower() == 'q':
            return False

    # Check COL 
    for c in range(COL):
        if c != col and board[row][c].lower() == 'q':
            return False 

    # Check Upper Left Diagonal
    for r, c in zip(range(row-1, -1, -1), range(col-1, -1, -1)):
        if board[r][c].lower() == 'q':
            return False
    
    # Check Lower Left Diagonal 
    for r, c in zip(range(row+1, ROW), range(col-1, -1, -1)):
        if board[r][c].lower() == 'q':
            return False

    # Check Upper Right Diagonal
    for r, c in zip(range(row-1, -1, -1), range(col+1, COL)):
        if board[r][c].lower() == 'q':
            return False

    # Check Lower Right Diagonal
    for r, c in zip(range(row+1, ROW), range(col+1, COL)):
        if board[r][c].lower() == 'q':
            return False
    return True


def backtrack_NQueens(board: list, queens: int, col: int):
    if queens == 0 or col == len(board):
        temp = []
        for row in board:
            temp.append([x for x in row])
        solutions.append(temp)
        return

    for r in range(len(board)):
        if board[r][col] == "Q":
            if isSafe((r,col), board):
                backtrack_NQueens(board, queens-1, col+1)
            return 

    for r in range(len(board)):
        if board[r][col] == '.' and isSafe((r, col), board):

            board[r][col] = 'q'
            backtrack_NQueens(board, queens-1, col+1)
            board[r][col] = '.'
            queens += 1 


def display_board(board: list):
    """Prints the Chess Board"""
    print(f"   {' '.join([str(i) for i in range(len(board))])}")
    print(" ","_"*(2*len(board)+1))
    for i, row in enumerate(board):
        print(f"{i} |{'|'.join(row)}|")
    print(" ","‾"*(2*len(board)+1))


if __name__ == "__main__":
    global solutions
    solutions = []
    N = int(input("Please Enter the Number of Queens to be placed: "))
    board = [['.']*N for _ in range(N)]
    display_board((board))

    if N > 0:
        while(True):
            x, y = map(int, input("Please Enter the position for your Queen (x y): ").split())
            board[x][y] = "Q"
            display_board(board)
            if (input("Are you happy with your choice? (y/N): ").lower()[0] == 'y'):
                break 
            else:
                board[x][y] = "."


        backtrack_NQueens(board, N, 0)
        print(f"{len(solutions)} were found for the given choice")
        if len(solutions) != 0:
            for i, sol in enumerate(solutions, 1):
                print(f"Solution :{i}")
                display_board(sol)



